fmpz/aors_ui.c

```diff
@@ -0,0 +1,338 @@
+        mf->_mp_size = -mf->_mp_size;
+    }
+}
+
+/* Subtract two limbs (they have different sign) and decide the sign via g. */
+static void
+_fmpz_sub_mpn_1(fmpz_t f, const mp_limb_t * glimbs, mp_size_t gsz, mp_limb_t x)
+{
+    __mpz_struct * mf;
+    mp_limb_t * flimbs;
+    mp_size_t gabssz = FLINT_ABS(gsz);
+
+    /* If size of g is 1, we have a higher probability of the result being
+     * small. */
+    if (gabssz == 1)
+    {
+        if (x <= glimbs[0]) /* Result is zero or has the same sign as g */
+        {
+            x = glimbs[0] - x;
+L1:         if (x <= COEFF_MAX) /* Fits in small fmpz */
+            {
+                if (COEFF_IS_MPZ(*f))
+                    _fmpz_clear_mpz(*f);
+                *f = (gsz > 0) ? x : -x; /* With consideration of sign */
+            }
+            else /* Does not fit in small fmpz */
+            {
+                if (COEFF_IS_MPZ(*f))
+                    mf = COEFF_TO_PTR(*f);
+                else
+                {
+                    mf = _fmpz_new_mpz();
+                    *f = PTR_TO_COEFF(mf);
+                }
+                mf->_mp_d[0] = x;
+                mf->_mp_size = gsz; /* Sign of f is the same as for g */
+            }
+        }
+        else /* |x| > |g|, which implies f has opposite sign of g */
+        {
+            /* Set x to the absolute value of |g - x|. By switching sign of
+             * gsz, we can reuse the code above. */
+            x -= glimbs[0];
+            gsz = -gsz;
+            goto L1;
+        }
+        return;
+    }
+
+    /* As g has more than one limb, it is a very high probability that result
+     * does not fit inside small fmpz. */
+    if (COEFF_IS_MPZ(*f))
+        mf = COEFF_TO_PTR(*f);
+    else
+    {
+        mf = _fmpz_new_mpz();
+        *f = PTR_TO_COEFF(mf);
+    }
+    flimbs = mf->_mp_d;
+
+    if (gabssz == 2)
+    {
+        /* Special case. Can result in a small fmpz, but as |g| > |x| the sign
+         * cannot change. */
+        sub_ddmmss(flimbs[1], flimbs[0], glimbs[1], glimbs[0], 0, x);
+        if (flimbs[1] != 0)
+        {
+            /* Most likely: upper limb not zero, so we just have set the sign
+             * of f to g's. */
+            mf->_mp_size = gsz;
+        }
+        else if (flimbs[0] > COEFF_MAX)
+        {
+            /* Still very likely: Upper limb is zero but lower limb does not
+             * fit inside a small fmpz. Sign is the same as for g, but the
+             * absolute value of the size is one. */
+            mf->_mp_size = (gsz > 0) ? 1 : -1;
+        }
+        else
+        {
+            /* Upper limb is zero and lower limb fits inside a small fmpz.
+             * Therefore we set f to +/- flimbs[0] and clear the mpz associated
+             * to f. */
+            slong tmp = flimbs[0]; /* We will clear this mpz, so save first. */
+            _fmpz_clear_mpz(*f);
+            *f = (gsz > 0) ? tmp : -tmp;
+        }
```

Comment:
Rarely gets tested. Had to take 100 times more tests in order to produce more than one triggering this. It passes, though.